src/modules/CSADigitizer/CSADigitizerModule.cpp

```diff
@@ -0,0 +1,337 @@
+    std::vector<PixelHit> hits;
+    for(auto& pixel_charge : pixel_message_->getData()) {
+        auto pixel = pixel_charge.getPixel();
+        auto pixel_index = pixel.getIndex();
+        auto inputcharge = static_cast<double>(pixel_charge.getCharge());
+
+        LOG(DEBUG) << "Received pixel " << pixel_index << ", charge " << Units::display(inputcharge, "e");
+
+	const auto& pulse = pixel_charge.getPulse(); // the pulse containing charges and times
+	auto pulse_vec = pulse.getPulse();           // the vector of the charges 
+        auto timestep = pulse.getBinning();
+	const int npx = static_cast<int>(ceil(tmax_/timestep));
+
+	if (first_event_){ // initialize impulse response function - assume all time bins are equal
+	  impulseResponse_ = new double[npx];
+	  for (int ipx=0; ipx<npx; ++ipx){
+	    impulseResponse_[ipx]=fImpulseResponse_->Eval(ipx*timestep);
+	  }
+	  first_event_ = false;
+	  LOG(TRACE) << "impulse response initalised. timestep  : " <<  timestep << ", tmax_ : " <<  tmax_ << ", npx " << npx;
+	}
+	
+	Pulse output_pulse(timestep);
+	auto input_length = pulse_vec.size();
+	LOG(TRACE) << "Preparing pulse for pixel " << pixel_index << ", " << pulse_vec.size() << " bins of "
+		   << Units::display(timestep, {"ps", "ns"})
+		   << ", total charge: " << Units::display(pulse.getCharge(), "e");
+	//convolution of the pulse (size input_length) with the impulse response (size npx)
+	for(unsigned long int k=0; k<npx; ++k){
+	  for(unsigned long int i=0; i<=k; ++i){
+	    if( (k-i) < input_length){
+	      //asv to do: not a charge, but voltage pulse... still use this?
+	      output_pulse.addCharge(pulse_vec.at(k-i) * impulseResponse_[i], timestep * static_cast<double>(k));
+	    }
+	  }
+	}
+	auto output_vec = output_pulse.getPulse(); 
+	LOG(TRACE) << "amplified signal" << output_pulse.getCharge()/1e9
+		   << " mV in a pulse with " << output_vec.size() << "bins";
+
+
+	
+	//TOA and TOT logic  
+	auto threshold = config_.get<double>("threshold");
+	bool is_over_threshold=false;
+	double toa, tot;
+	for(long unsigned int i=0; i<output_vec.size(); ++i){
+
+	  //use while instead of for maybe?
+	  // set time of arrival when pulse first crosses threshold, start countint ToT
+	  if (output_vec.at(i)>threshold && !is_over_threshold){
+	    is_over_threshold = true;
+	    toa= i * timestep;
+	    tot=timestep;
+	    LOG(TRACE) << "now starting the tot counting...";
+	  };
+	  // asv need to add some precision to avoid noise stopping tot
+	  // or do TOT/TOA before adding noise?
+	  if (is_over_threshold){
+	    if (output_vec.at(i)>threshold){	      
+	      tot+=timestep;
+	    }
+	    else{
+	      is_over_threshold=false;
+	      break;
+	    }
+	  }
+	  else if (output_vec.size()-1 == i){ // pulse over, never crossed the threshold
+	    toa=0;
+	    tot=0;
+	    LOG(TRACE) << " - never crossing the threshold";
+	  }
+	}
+ 	LOG(TRACE) << "ToA " << toa << " ns, ToT " << tot << "ns";
+
+	
+
+	//apply noise on the amplified pulse
```

Comment:
This should go before we evaluate things from this pulse, at this point of the module we are done with the simulation and the only thing we continue doing is some debug plotting

(we don"t consider any of the histograms produced by modules directly a simulation result but merely additional information for debugging proposes)